pgmpy/base/MAG.py

```diff
@@ -0,0 +1,295 @@
+    """
+
+    def __init__(
+        self,
+        ebunch: Optional[Iterable[tuple[Hashable, Hashable]]] = None,
+        latents: set[Hashable] = set(),
+        roles=None,
+    ):
+        """
+        Initialize a Maximal Ancestral Graph
+
+        Parameters
+        ----------
+        ebunch : iterable of tuples, optional
+            A list or iterable of edges to add at initialization.
+
+        latents : set, default=set()
+            Set of latent (unobserved) variables.
+
+        roles : dict, optional (default: None)
+            A dictionary mapping roles to node names.
+            The keys are roles, and the values are role names (strings or iterables of str).
+            If provided, this will automatically assign roles to the nodes in the graph.
+            Passing a key-value pair via ``roles`` is equivalent to calling
+            ``with_role(role, variables)`` for each key-value pair in the dictionary.
+
+        Returns
+        -------
+        MAG
+            A new instance of a Maximal Ancestral Graph.
+
+        Examples
+        --------
+        >>> from pgmpy.base import MAG
+        >>> mag = MAG(
+        ...     ebunch=[("L", "A", "-", ">"), ("B", "C", "-", ">")], latents={"L"}
+        ... )
+        >>> sorted(mag.nodes())
+        ['A', 'B', 'C', 'L']
+
+        Roles can be assigned to nodes in the graph at construction or using methods.
+
+        At construction:
+
+        >>> mag = MAG(
+        ...     ebunch=[("L", "A", "-", ">"), ("B", "C", "-", ">")],
+        ...     latents={"L"},
+        ...     roles={"exposure": "A", "outcome": "B"},
+        ... )
+
+        Roles can also be assigned after creation using ``with_role`` method.
+
+        >>> mag = mag.with_role("adjustment", {"L", "C"})
+
+        Vertices of a specific role can be retrieved using ``get_role`` method.
+
+        >>> mag.get_role("exposure")
+        ["A"]
+        >>> mag.get_role("adjustment")
+        ["L", "C"]
+
+        """
+        super().__init__(ebunch=ebunch, latents=latents, roles=roles)
+
+    def _is_collider(self, u, c, v):
+        """
+        Check if a node is a collider in a path u - c - v.
+
+        A collider is a node with incoming arrowheads on both sides:
+        u -> c <- v.
+
+        Parameters
+        ----------
+        u : Hashable
+            The first endpoint in the triple (u, c, v).
+        c : Hashable
+            The middle node, candidate collider.
+        v : Hashable
+            The second endpoint in the triple.
+
+        Returns
+        -------
+        bool
+            True if `c` is a collider on the path, False otherwise.
+
+        Examples
+        --------
+        >>> from pgmpy.base import MAG
+        >>> mag = MAG()
+        >>> mag.add_edge("X", "Z", "-", ">")
+        >>> mag.add_edge("Y", "Z", "-", ">")
+        >>> mag._is_collider("X", "Z", "Y")
+        True
+        """
+        mark_uc_at_c = self.edges[u, c]["marks"][c]
```

Comment:
We will need to add checks to see whether the edges exist. Without the checks, the current code would throw a KeyError, which would be hard to debug.